In [10]:
import gzip
import numpy as np
from google.colab import drive
drive.mount('/content/drive')
import gzip
import numpy as np

def load_mnist_images(filename):
    with gzip.open(filename, 'rb') as f:
        data = np.frombuffer(f.read(), np.uint8, offset=16)
    return data.reshape(-1, 28*28)

def load_mnist_labels(filename):
    with gzip.open(filename, 'rb') as f:
        labels = np.frombuffer(f.read(), np.uint8, offset=8)
    return labels

train_data_path = 'drive/My Drive/zhang/train-images-idx3-ubyte.gz'
train_labels_path = 'drive/My Drive/zhang/train-labels-idx1-ubyte.gz'
test_data_path = 'drive/My Drive/zhang/t10k-images-idx3-ubyte.gz'
test_labels_path = 'drive/My Drive/zhang/t10k-labels-idx1-ubyte.gz'

X_train = load_mnist_images(train_data_path) / 255.
y_train = load_mnist_labels(train_labels_path)
X_test = load_mnist_images(test_data_path) / 255.
y_test = load_mnist_labels(test_labels_path)

def softmax(X):
    expX = np.exp(X - np.max(X))
    return expX / expX.sum(axis=1, keepdims=True)

def compute_cost(X, y, theta, num_labels):
    m = X.shape[0]
    predictions = softmax(np.dot(X, theta))
    y_onehot = np.eye(num_labels)[y]
    logprobs = -np.log(predictions[range(m), y])
    cost = np.sum(logprobs) / m
    grad = (1/m) * np.dot(X.T, predictions - y_onehot)
    return cost, grad

def gradient_descent(X, y, theta, alpha, num_iterations, num_labels):
    cost_history = []
    for i in range(num_iterations):
        cost, grad = compute_cost(X, y, theta, num_labels)
        theta -= alpha * grad
        cost_history.append(cost)
    return theta, cost_history

m = X_train.shape[0]
X_train_bias = np.hstack((np.ones((m, 1)), X_train))
num_labels = 10
theta = np.zeros((X_train_bias.shape[1], num_labels))
theta, cost_history = gradient_descent(X_train_bias, y_train, theta, 0.001, 1200, num_labels)


def predict(X, theta):
    probs = softmax(np.dot(X, theta))
    return np.argmax(probs, axis=1)

def confusion_matrix(y_true, y_pred):
    num_classes = len(np.unique(y_true))
    cm = np.zeros((num_classes, num_classes), dtype=int)
    for i in range(len(y_true)):
        cm[y_true[i]][y_pred[i]] += 1
    return cm

def metrics_from_cm(cm):
    tn = np.diag(cm)
    fp = cm.sum(axis=0) - tn
    fn = cm.sum(axis=1) - tn
    tp = tn

    accuracy = tp.sum() / cm.sum()
    sensitivity = tp / (tp + fn)
    specificity = tp / (tp + fp)
    f1_score = 2*tp / (2*tp + fp + fn)

    return accuracy, sensitivity, specificity, f1_score

m_test = X_test.shape[0]
X_test_bias = np.hstack((np.ones((m_test, 1)), X_test))
y_pred = predict(X_test_bias, theta)
cm = confusion_matrix(y_test, y_pred)
accuracy, sensitivity, specificity, f1 = metrics_from_cm(cm)

print("Accuracy:", accuracy)
print("Sensitivity:", sensitivity)
print("Specificity:", specificity)
print("F1 Score:", f1)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Accuracy: 0.6595
Sensitivity: [0.819 0.902 0.62  0.845 0.651 0.007 0.029 0.881 0.9   0.941]
Specificity: [0.67630058 0.96989247 0.54577465 0.71008403 0.46202981 1.
 0.43283582 0.60466712 0.85959885 0.60866753]
F1 Score: [0.74084125 0.93471503 0.58052434 0.7716895  0.54047323 0.01390268
 0.05435801 0.71713472 0.87933561 0.73919874]


Data Handling
Where Data Came From: I took the dataset from Google Drive.
How I Loaded Data: I used Python's gzip module and Numpy to get the MNIST Fashion images and labels.
Data Look: I changed the images into 784 dimensions (28x28) lines. The labels have 10 types.
Which Model I Used
Kind of Model: I picked a simple line method for this many-type classification problem.
How I Did It: I used Softmax Regression because it's good for many-type classification.
How I Made the Model Learn
Which Loss Function: I used Cross-entropy loss.
How I Improved It: I made it better with Gradient Descent.
How Many Times I Tried: I tried 1200 times.
How Fast I Made Changes: I set the speed at 0.001.
How the Model Did
How Right It Was: The model did okay. It got right 0.6595 times out of 1.
Sensitivity: For some types like 5 and 6, the model didn't do well.
Specificity: For type 5, the model did very well. It didn't say any wrong ones were right.
F1 Score: Like Sensitivity, the model didn't do well for some types.
My Thoughts
The line method, Softmax Regression maybe it's not the best for complicated data like MNIST Fashion.
Maybe using more complicated models like CNNs would be better for MNIST Fashion.
In the future, I can try changing some settings or use better ways to prepare the data to make the model do better.





For the MNIST Fashion dataset, Softmax Regression might not be the best choice. Maybe using something like Convolutional Neural Networks (CNNs) is better. In the future, I think I can try to change some settings, maybe add some other techniques, or prepare the data in different ways to make the results better.